File: AkhterAirPlayReceiver/app/src/main/java/com/cjx/airplayjavademo/player/VideoPlayer.java
package com.cjx.airplayjavademo.player;

import android.annotation.SuppressLint;
import android.media.MediaCodec;
import android.media.MediaFormat;
import android.os.Handler;
import android.os.HandlerThread;
import android.util.Log;
import android.view.Surface;


import com.cjx.airplayjavademo.model.NALPacket;
import com.cjx.airplayjavademo.tools.LogRepository;

import java.nio.ByteBuffer;
import java.util.Objects;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class VideoPlayer {
    private static final String TAG = "VideoPlayer";
    private static final String MIME_TYPE = MediaFormat.MIMETYPE_VIDEO_AVC;
    private final int mVideoWidth = 540;
    private final int mVideoHeight = 960;
    private final MediaCodec.BufferInfo mBufferInfo = new MediaCodec.BufferInfo();
    private MediaCodec mDecoder = null;
    private final Surface mSurface;
    private BlockingQueue<NALPacket> packets = new LinkedBlockingQueue<>(500);
    private final HandlerThread mDecodeThread = new HandlerThread("VideoDecoder");

    private final MediaCodec.Callback mDecoderCallback = new MediaCodec.Callback() {
        @Override
        public void onInputBufferAvailable(MediaCodec codec, int index) {
            try {
                NALPacket packet = packets.take();
                ByteBuffer inputBuffer = codec.getInputBuffer(index);
                if (inputBuffer != null) {
                    inputBuffer.put(packet.nalData);
                    codec.queueInputBuffer(index, 0, packet.nalData.length, packet.pts, 0);
                }
            } catch (InterruptedException e) {
                Log.e(TAG, "Error while waiting for NALPacket", e);
            } catch (IllegalStateException e) {
                Log.e(TAG, "Error while queuing input buffer", e);
            }
        }

        @Override
        public void onOutputBufferAvailable(MediaCodec codec, int index, MediaCodec.BufferInfo info) {
            try {
                // Rilascia il buffer all'output Surface
                codec.releaseOutputBuffer(index, true);
                LogRepository.INSTANCE.addLog(TAG, "Released output buffer index=" + index, 'I');
            } catch (IllegalStateException e) {
                Log.e(TAG, "Error releasing output buffer", e);
            }
        }

        @Override
        public void onError(MediaCodec codec, MediaCodec.CodecException e) {
            Log.e(TAG, "Decoder error", e);
        }

        @Override
        public void onOutputFormatChanged(MediaCodec codec, MediaFormat format) {
            Log.i(TAG, "Output format changed to " + format);
        }
    };

    public VideoPlayer(Surface surface, int width, int heigth) {
//        this.mVideoWidth=width;
//        this.mVideoHeight=heigth;
        mSurface = surface;
    }

    public void initDecoder() {
        mDecodeThread.start();
        try {
            // 解码分辨率
            Log.i(TAG, "initDecoder: mVideoWidth=" + mVideoWidth + "---mVideoHeight=" + mVideoHeight);
            mDecoder = MediaCodec.createDecoderByType(MIME_TYPE);
            MediaFormat format = MediaFormat.createVideoFormat(MIME_TYPE, mVideoWidth, mVideoHeight);
            mDecoder.configure(format, mSurface, null, 0);
            mDecoder.setVideoScalingMode(MediaCodec.VIDEO_SCALING_MODE_SCALE_TO_FIT);
            mDecoder.setCallback(mDecoderCallback, new Handler(mDecodeThread.getLooper()));
            mDecoder.start();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void addPacket(NALPacket nalPacket) {
        try {
            packets.put(nalPacket);
        } catch (InterruptedException e) {
            // 队列满了
            Log.e(TAG, "run: put error:", e);
        }
    }

    public void start() {
        initDecoder();
    }


    public void release() {
        if (mDecoder != null) {
            try {
                mDecoder.stop();
                mDecoder.release();
            } catch (Exception e) {
                Log.e(TAG, "Error while releasing MediaCodec", e);
            }
        }

        if (mDecodeThread != null && mDecodeThread.isAlive()) {
            mDecodeThread.quitSafely();
        }

        packets.clear();  // Rilascia il buffer dei pacchetti

        Log.d(TAG, "VideoPlayer resources released.");

    }

    public void stopVideoPlay() {
        try {
            mDecoder.stop();
        } catch (Exception e) {
            e.printStackTrace();
        }
        try {
            mDecoder.release();
        } catch (Exception e) {
            e.printStackTrace();
        }
        mDecodeThread.quit();
        packets.clear();
    }

    public void stopPlayer() {
        if (mDecoder != null) {
            try {
                mDecoder.stop();
                mDecoder.release();
            } catch (Exception e) {
                Log.e(TAG, "Errore durante il rilascio del MediaCodec", e);
                LogRepository.INSTANCE.addLog(TAG, "Errore durante il rilascio del MediaCodec", 'E');
            }
        }

        if (mDecodeThread != null && mDecodeThread.isAlive()) {
            mDecodeThread.quitSafely();
            try {
                mDecodeThread.join(); // Assicura che il thread si sia fermato
                LogRepository.INSTANCE.addLog(TAG, "VideoPlayer thread stopped.", 'W');
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                Log.e(TAG, "Errore durante l'interruzione di mDecodeThread", e);
            }
        }

        packets.clear();
        Log.d(TAG, "Risorse VideoPlayer rilasciate.");
        LogRepository.INSTANCE.addLog(TAG, "Risorse VideoPlayer rilasciate.", 'I');
    }

}

File: AkhterAirPlayReceiver/app/src/main/java/com/cjx/airplayjavademo/player/VideoPlayer.kt
package com.cjx.airplayjavademo.player

import android.media.MediaCodec
import android.media.MediaFormat
import android.os.Handler
import android.os.HandlerThread
import android.view.Surface
import com.cjx.airplayjavademo.model.NALPacket
import com.cjx.airplayjavademo.tools.LogRepository
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.atomic.AtomicBoolean

class VideoPlayer(private val mSurface: Surface) {
    companion object {
        private const val TAG = "VideoPlayer.kt"
        private const val MIME_TYPE = MediaFormat.MIMETYPE_VIDEO_AVC
        private const val VIDEO_WIDTH = 540
        private const val VIDEO_HEIGHT = 960
        private const val MAX_QUEUE_SIZE = 240
        private const val LOG_FIRST_PACKETS = 5
        private const val LOG_EVERY_PACKETS = 100
    }

    private var mDecoder: MediaCodec? = null
    private var mDecodeThread: HandlerThread? = null
    private val packets = LinkedBlockingQueue<NALPacket>(MAX_QUEUE_SIZE)
    private val availableInputBuffers = ConcurrentLinkedQueue<Int>()
    private val drainLock = Any()
    private val released = AtomicBoolean(false)
    private val failed = AtomicBoolean(false)

    private var queuedPackets = 0
    private var droppedPackets = 0
    private var inputCallbacks = 0
    private var queuedInputBuffers = 0
    private var outputBuffers = 0

    private val mDecoderCallback = object : MediaCodec.Callback() {
        override fun onInputBufferAvailable(codec: MediaCodec, index: Int) {
            inputCallbacks++
            if (shouldLogPacket(inputCallbacks)) {
                LogRepository.addLog(TAG, "Input buffer available index=$index queueSize=${packets.size}", 'I')
            }
            availableInputBuffers.offer(index)
            drainInputBuffers(codec)
        }

        override fun onOutputBufferAvailable(codec: MediaCodec, index: Int, info: MediaCodec.BufferInfo) {
            outputBuffers++
            val surfaceValid = mSurface.isValid
            val render = surfaceValid && !released.get() && !failed.get() && info.size > 0
            try {
                codec.releaseOutputBuffer(index, render)
                if (shouldLogPacket(outputBuffers)) {
                    LogRepository.addLog(TAG, "Output buffer released index=$index render=$render " +
                            "size=${info.size} ptsUs=${info.presentationTimeUs} flags=${info.flags} " +
                            "surfaceValid=$surfaceValid", 'I')
                }
                if (!surfaceValid) {
                    failDecoder("Surface became invalid during output release", null)
                }
            } catch (e: Exception) {
                failDecoder("Error releasing output buffer index=$index", e)
            }
        }

        override fun onError(codec: MediaCodec, e: MediaCodec.CodecException) {
            failDecoder("MediaCodec error diagnostic=${e.diagnosticInfo} recoverable=${e.isRecoverable} transient=${e.isTransient}", e)
        }

        override fun onOutputFormatChanged(codec: MediaCodec, format: MediaFormat) {
            LogRepository.addLog(TAG, "Output format changed to $format", 'I')
        }
    }

    fun initDecoder() {
        if (!mSurface.isValid) {
            failDecoder("Cannot initialize decoder: Surface is invalid", null)
            return
        }

        val decodeThread = HandlerThread("VideoDecoder").also {
            it.start()
            mDecodeThread = it
        }

        try {
            LogRepository.addLog(TAG, "initDecoder: width=$VIDEO_WIDTH height=$VIDEO_HEIGHT surfaceValid=${mSurface.isValid}", 'I')
            val decoder = MediaCodec.createDecoderByType(MIME_TYPE)
            mDecoder = decoder
            val format = MediaFormat.createVideoFormat(MIME_TYPE, VIDEO_WIDTH, VIDEO_HEIGHT)
            decoder.setCallback(mDecoderCallback, Handler(decodeThread.looper))
            decoder.configure(format, mSurface, null, 0)
            decoder.setVideoScalingMode(MediaCodec.VIDEO_SCALING_MODE_SCALE_TO_FIT)
            decoder.start()
            LogRepository.addLog(TAG, "MediaCodec decoder started.", 'I')
        } catch (e: Exception) {
            failDecoder("Error while initializing MediaCodec", e)
        }
    }

    fun addPacket(nalPacket: NALPacket) {
        val data = nalPacket.nalData
        if (released.get() || failed.get() || data == null) {
            droppedPackets++
            logDrop("Dropping video packet because decoder is unavailable or packet is empty")
            return
        }

        if (!mSurface.isValid) {
            droppedPackets++
            failDecoder("Surface invalid while adding video packet seq=${nalPacket.sequenceNumber}", null)
            return
        }

        if (!packets.offer(nalPacket)) {
            packets.poll()
            droppedPackets++
            logDrop("Dropping oldest video packet because decoder queue is full")
            if (!packets.offer(nalPacket)) {
                droppedPackets++
                logDrop("Dropping incoming video packet because decoder queue remains full")
                return
            }
        }

        queuedPackets++
        if (shouldLogPacket(queuedPackets)) {
            LogRepository.addLog(TAG, "Queued video packet seq=${nalPacket.sequenceNumber} " +
                    "ptsUs=${nalPacket.pts} bytes=${data.size} codecConfig=${nalPacket.codecConfig} " +
                    "queueSize=${packets.size}", 'I')
        }

        mDecoder?.let { drainInputBuffers(it) }
    }

    fun start() {
        initDecoder()
    }

    fun isAvailable(): Boolean {
        return !released.get() && !failed.get() && mDecoder != null && mSurface.isValid
    }

    fun release() {
        releaseInternal("release()", true)
    }

    fun stopPlayer() {
        releaseInternal("stopPlayer()", true)
    }

    private fun drainInputBuffers(codec: MediaCodec) {
        synchronized(drainLock) {
            if (released.get() || failed.get()) {
                return
            }

            while (true) {
                val inputIndex = availableInputBuffers.peek() ?: return
                val packet = packets.poll() ?: return
                availableInputBuffers.poll()

                val data = packet.nalData
                if (data == null || data.isEmpty()) {
                    droppedPackets++
                    logDrop("Dropping empty video packet before decoder input")
                    continue
                }

                try {
                    val inputBuffer = codec.getInputBuffer(inputIndex)
                    if (inputBuffer == null) {
                        droppedPackets++
                        LogRepository.addLog(TAG, "Input buffer $inputIndex is null; dropping seq=${packet.sequenceNumber}", 'E')
                        continue
                    }

                    inputBuffer.clear()
                    if (data.size > inputBuffer.capacity()) {
                        failDecoder("NAL packet larger than input buffer. seq=${packet.sequenceNumber} " +
                                "packetBytes=${data.size} capacity=${inputBuffer.capacity()}", null)
                        return
                    }

                    inputBuffer.put(data)
                    val flags = if (packet.codecConfig) MediaCodec.BUFFER_FLAG_CODEC_CONFIG else 0
                    codec.queueInputBuffer(inputIndex, 0, data.size, packet.pts, flags)
                    queuedInputBuffers++
                    if (shouldLogPacket(queuedInputBuffers)) {
                        LogRepository.addLog(TAG, "Queued decoder input index=$inputIndex seq=${packet.sequenceNumber} " +
                                "ptsUs=${packet.pts} bytes=${data.size} flags=$flags queueSize=${packets.size}", 'I')
                    }
                } catch (e: Exception) {
                    failDecoder("Error queuing decoder input index=$inputIndex seq=${packet.sequenceNumber}", e)
                    return
                }
            }
        }
    }

    private fun failDecoder(reason: String, throwable: Throwable?) {
        if (!failed.compareAndSet(false, true)) {
            return
        }
        if (throwable != null) {
            LogRepository.addLog(TAG, "$reason: ${throwable.message}", 'E')
        } else {
            LogRepository.addLog(TAG, reason, 'E')
        }

        Thread {
            releaseInternal("decoder failure: $reason", true)
        }.start()
    }

    private fun releaseInternal(reason: String, joinThread: Boolean) {
        if (!released.compareAndSet(false, true)) {
            return
        }

        LogRepository.addLog(TAG, "Releasing VideoPlayer resources. reason=$reason " +
                "queueSize=${packets.size} availableInputs=${availableInputBuffers.size}", 'I')

        packets.clear()
        availableInputBuffers.clear()

        val decoder = mDecoder
        mDecoder = null
        if (decoder != null) {
            try {
                decoder.stop()
                LogRepository.addLog(TAG, "MediaCodec stopped.", 'I')
            } catch (e: Exception) {
                LogRepository.addLog(TAG, "MediaCodec stop failed: ${e.message}", 'W')
            }
            try {
                decoder.release()
                LogRepository.addLog(TAG, "MediaCodec released.", 'I')
            } catch (e: Exception) {
                LogRepository.addLog(TAG, "MediaCodec release failed: ${e.message}", 'E')
            }
        }

        val decodeThread = mDecodeThread
        mDecodeThread = null
        if (decodeThread != null && decodeThread.isAlive) {
            decodeThread.quitSafely()
            if (joinThread && Thread.currentThread() != decodeThread) {
                try {
                    decodeThread.join(1000)
                    LogRepository.addLog(TAG, "Video decoder thread stopped. alive=${decodeThread.isAlive}", 'I')
                } catch (e: InterruptedException) {
                    Thread.currentThread().interrupt()
                    LogRepository.addLog(TAG, "Interrupted while stopping decoder thread", 'E')
                }
            }
        }

        LogRepository.addLog(TAG, "VideoPlayer resources released.", 'I')
    }

    private fun logDrop(reason: String) {
        if (shouldLogPacket(droppedPackets)) {
            LogRepository.addLog(TAG, "$reason dropped=$droppedPackets queueSize=${packets.size}", 'W')
        }
    }

    private fun shouldLogPacket(count: Int): Boolean {
        return count <= LOG_FIRST_PACKETS || count % LOG_EVERY_PACKETS == 0
    }
}
